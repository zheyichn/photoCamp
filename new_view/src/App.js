// eslint-disable react/react-in-jsx-scope
import "./App.css";
import CreatePostPage from "./components/Pages/CreatePostPage";
import FriendsSuggestionsPage from "./components/Pages/FriendsSuggestionPage";
import ShowAllUserPage from "./components/Pages/ShowAllUserPage";
import HomePage from "./components/Pages/HomePage";
import LandingPage from "./components/Pages/LandingPage";
import ProfilePage from "./components/Pages/ProfilePage";
import FollowerPage from "./components/Pages/FollowerPage";
import FollowingPage from "./components/Pages/FollowingPage";
// import Header from "./components/Activity/Header";
import CreateNewAccountPage from "./components/Pages/CreateNewAccountPage";
import ForgotPasswordPage from "./components/Pages/ForgotPasswordPage";
import {
  BrowserRouter as Router,
  Routes,
  Route,
  Navigate,
} from "react-router-dom";
// import { Redirect } from "react-router-dom";
import { createContext } from "react";

export const Context = createContext();
// require("dotenv").config();

function App() {
  const loggedInUser = {
    userName: "admin",
    profile:
      "https://user-images.githubusercontent.com/93358071/191072941-5fc4a59a-49d0-4b18-80f9-573e38984fa8.jpeg",
    userId: 1,
  };

  return (
    <Context.Provider value={loggedInUser}>
      <div>
        <Router>
          <Routes>
            <Route path="/" element={<LandingPage />} />
            <Route
              path="/createnewaccount"
              element={<CreateNewAccountPage />}
            />
            <Route path="/forgotpassword" element={<ForgotPasswordPage />} />
            {sessionStorage.getItem("app-token") !== null ? (
              <>
                <Route path="/homepage" element={<HomePage />} />
                <Route path="/post" element={<CreatePostPage />} />
                <Route
                  path="/suggestion"
                  element={<FriendsSuggestionsPage />}
                />
                <Route path="/profile" element={<ProfilePage />} />
                <Route path="/display" element={<ShowAllUserPage />} />
                <Route path="/following" element={<FollowingPage />} />
                <Route path="/follower" element={<FollowerPage />} />
              </>
            ) : null}
            {/* catch all: back to root page if not logged in */}
            <Route path="*" element={<Navigate to="/" />} />
          </Routes>
        </Router>
      </div>
    </Context.Provider>
  );
}

export default App;
