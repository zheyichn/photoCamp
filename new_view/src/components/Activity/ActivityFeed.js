/* eslint-disable no-unused-vars */
import React, { useState, useRef, useEffect } from "react";
import {
  Alert,
  Avatar,
  Button,
  Card,
  CardHeader,
  CardMedia,
  CardContent,
  Grid,
} from "@mui/material";
import CardActions from "@mui/material/CardActions";
import Collapse from "@mui/material/Collapse";
import IconButton, { IconButtonProps } from "@mui/material/IconButton";
import FavoriteIcon from "@mui/icons-material/Favorite";
import ChatBubbleOutlineRoundedIcon from "@mui/icons-material/ChatBubbleOutlineRounded";
import CommentForm from "./CommentForm";
import Comment from "./Comment";
import EditableComment from "./EditableComment";
import UserDialog from "./UserDialog";
import { isImgUrl } from "../../utility/isImageUrl";
import {
  addLike,
  addNewComment,
  getCommentorById,
  getCommentsByPostId,
  hidePost,
  unLike,
} from "../../api/mock_api_k";
import LinearProgress from "@mui/material/LinearProgress";
import VisibilityOffIcon from "@mui/icons-material/VisibilityOff";
import {
  getUserAPI,
} from "../../api/mock_api_following_post_follower";

export default function ActivityFeed(props) {
  const post = props.post;
  const {
    _id,
    creatorName,
    postContent,
    caption,
    creatorProfile,
    // these props are left for future feature implementations
    // visibility,
    // postTime,
    taggedUsers,
    likedByUsers = [],
  } = post;

  // const user = useContext(Context);
  // const userId = sessionStorage.getItem("userId");
  const [expanded, setExpanded] = useState(false);
  const [commentsList, setCommentsList] = useState([]);
  const [error, setError] = useState("");
  const newComment = useRef("");
  const likedByCurrentUser = likedByUsers.includes(
    sessionStorage.getItem("userId")
  ); // for UI testing, since testing file cannot use useContext, by default pass a loggedin userId as 1
  const [liked, setLiked] = useState(likedByCurrentUser);
  const [isPosting, setIsPosting] = useState(false);
  const [isImage, setIsImage] = useState(true);
  const [isHidden, setIsHidden] = useState(false);
  // for user dialog
  const [dialogOpen, setDialogOpen] = useState(false);
  const [dialogUser, setDialogUser] = useState("");
  const [dialogUserInfo, setDialogUserInfo] = useState({});

  // handle the opening of user dialog
  const handleDialogOpen = async(user) => {
    setDialogOpen(true);
    setDialogUser(user);
    const results = await getUserAPI(user);
    setDialogUserInfo(results.data);
  };

  // handle the closing of user dialog
  const handleDialogClose = () => {
    setDialogOpen(false);
    setDialogUser("");
  };

  const judgeIfImage = async (url) => {
    const res = await isImgUrl(url);
    setIsImage(res);
  };

  useEffect(() => {
    judgeIfImage(postContent);
  }, [postContent]);

  const fetchComments = async (postId) => {
    let comments;
    try {
      //  this is a array of comments object that are already sorted by time in descending order
      comments = await getCommentsByPostId(postId);
      // as a comment object doesn't has userProfile and userName information, we need to get userById
      for (let i = 0; i < comments.length; i++) {
        const commentorId = comments[i].commentorId;
        const commentor = await getCommentorById(commentorId);
        comments[i]["commentorName"] = commentor.commentorName;
        comments[i]["commentorProfile"] = commentor.commentorProfile;
      }
      setCommentsList(comments);
    } catch (e) {
      console.error(comments);
      console.error(e);
    }
  };

  const handleWriteCommentClick = async () => {
    await fetchComments(_id);
    setExpanded(!expanded);
  };

  const handleLikeClick = async () => {
    if (!liked) {
      // if this post is not liked by the current user, addlike
      addLike(_id, sessionStorage.getItem("userId"));
    } else {
      // if already been liked by the user, unlike the post
      unLike(_id, sessionStorage.getItem("userId"));
    }
    setLiked(!liked);
  };

  const handleHidePostClick = async () => {
    try {
      await hidePost(_id, sessionStorage.getItem("userName"));
      setIsHidden(true);
    } catch (e) {
      console.error(e);
    }
  };

  const likeIconColor = liked ? "red" : "grey";

  if (isHidden) {
    return null;
  }

  return (
    <Grid m={2}>
      <Card>
        <CardHeader
          avatar={<Avatar src={creatorProfile}></Avatar>}
          title={creatorName}
        ></CardHeader>

        {isImage ? (
          <CardMedia
            data-testid="test-img"
            component="img"
            height="300px"
            src={postContent}
            alt="Post Picture"
          ></CardMedia>
        ) : (
          <CardMedia
            data-testid="test-img"
            component="video"
            height="300px"
            src={postContent}
            autoPlay
            controls
            muted
            alt="Post Video"
          ></CardMedia>
        )}
        <CardContent data-testid="caption">{caption}</CardContent>
        <CardContent>
          <b>Tagged users:</b>
          {taggedUsers ? taggedUsers.map((user) => {
            return(
              <Button
                key={ _id + user }
                sx={{ textTransform: 'none' }}
                onClick={() => handleDialogOpen(user)}
              >
                { user + " "}
              </Button>
            )
          }) : null}
        </CardContent>
        <CardActions disableSpacing>
          <IconButton aria-label="like" onClick={handleLikeClick}>
            <FavoriteIcon style={{ color: likeIconColor }} />
          </IconButton>
          <IconButton
            className="openComment"
            aria-label="comment"
            data-testid="write-test"
            onClick={handleWriteCommentClick}
          >
            <ChatBubbleOutlineRoundedIcon />
          </IconButton>
          <IconButton
            aria-label="hide"
            data-testid="hide-test"
            onClick={handleHidePostClick}
          >
            <VisibilityOffIcon />
          </IconButton>
        </CardActions>
        <Collapse in={expanded} timeout="auto" unmountOnExit>
          <CardContent data-testid="comment-section">
            {commentsList.map((data) => {
              if (data.commentorId === sessionStorage.getItem("userId")) {
                return (
                  <div key={data._id}>
                    <EditableComment
                      createdAt={data.createdAt}
                      commentorName={data.commentorName}
                      commentText={data.commentText}
                      commentorProfile={data.commentorProfile}
                      postId={post._id}
                      commentId={data._id}
                      commentorId={data.commentorId}
                      reFetchComments={fetchComments}
                    />
                  </div>
                );
              } else {
                return (
                  <div key={data._id}>
                    <Comment
                      createdAt={data.createdAt}
                      commentorName={data.commentorName}
                      commentText={data.commentText}
                      commentorProfile={data.commentorProfile}
                    />
                  </div>
                );
              }
            })}
            <CommentForm
              data-testid="form-test"
              newComment={newComment}
              handlePostClick={async () => {
                if (newComment.current.value === "") {
                  setError("comments cannot be empty, try again!");
                } else {
                  const createdAt = Date.now();
                  // this ensures at FE, the userProfile who commments
                  // is always logged in users, not users randomly generated from
                  // mock.api. commentorProfile isn't an attribute in Comment's schema
                  // so I created myLocalComment purly for FE consistency
                  const myLocalComment = {
                    commentorName: sessionStorage.getItem("userName"),
                    commentorProfile: sessionStorage.getItem("userProfile"), // commentorProfile and name should be later on got from cookies
                    commentText: newComment.current.value,
                    commentorId: sessionStorage.getItem("userId"),
                    createdAt: createdAt,
                  };
                  const myComment = {
                    commentText: newComment.current.value,
                    commentorId: sessionStorage.getItem("userId"),
                    createdAt: createdAt,
                  };
                  try {
                    // only show the comment at the FE when the comment has been added to BE successfully
                    setIsPosting(true);
                    const newCommentId = await addNewComment(_id, myComment);
                    setIsPosting(false);
                    myLocalComment._id = newCommentId;
                    setCommentsList((prevComments) => {
                      return [...prevComments, myLocalComment];
                    });
                    setError("");
                    newComment.current.value = "";
                  } catch (e) {
                    console.log(e);
                    setIsPosting(false);
                    setError(`Please try again later. Encountered: ${e}`);
                  }
                }
              }}
            />
            {isPosting ? <LinearProgress /> : null}
            {error ? (
              <Grid mt={2}>
                <Alert severity="warning">{error}</Alert>
              </Grid>
            ) : null}
          </CardContent>
        </Collapse>
      </Card>
      <UserDialog
        dialogUser={dialogUser}
        dialogOpen={dialogOpen}
        handleDialogClose={handleDialogClose}
        dialogUserInfo={dialogUserInfo}
      />
    </Grid>
  );
}
