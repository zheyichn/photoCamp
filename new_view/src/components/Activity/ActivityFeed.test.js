import { render, screen, cleanup, within } from "@testing-library/react";
import "@testing-library/jest-dom";
import ActivityFeed from "./ActivityFeed";
import renderer from "react-test-renderer";
import userEvent from "@testing-library/user-event";
import { Context } from "../../App";

// const axios = require("axios");
// const MockAdapter = require("axios-mock-adapter");
// const mockAxios = new MockAdapter(axios);
// const commentsArray = [
//   { commentId: 1, commentText: "test comment one" },
//   { commentId: 2, commentText: "test comment two" },
//   { commentId: 3, commentText: "test comment three" },
// ];
// mockAxios.onGet().reply(200, commentsArray);
const fakePost = {
  creatorId: 10,
  creatorUsername: "Irwin",
  postContent:
    "https://user-images.githubusercontent.com/93358071/191072941-5fc4a59a-49d0-4b18-80f9-573e38984fa8.jpeg",
  caption:
    "Cum consectetur et. At quia sed non accusantium saepe qui cum. Sed est asperiores qui.",
  visibility: false,
  postTime: 1665776721,
  likedByUsers: [1],
  taggedUsers: [],
  creatorProfile: "http://loremflickr.com/640/480/technics",
  postId: "1",
};

const fakeLoggedInUser = {
  userName: "admin klaus",
  profile:
    "https://user-images.githubusercontent.com/93358071/191072941-5fc4a59a-49d0-4b18-80f9-573e38984fa8.jpeg",
  userId: 1,
};

// afterEach function runs after each test suite is executed
afterEach(() => {
  cleanup();
});

describe("ActivityFeed Component", () => {
  // Test1: sub components exixt
  test("ActivityFeed has postContent and button", () => {
    render(<ActivityFeed post={fakePost} />);
    const cardMedia = screen.getByTestId("test-img");
    expect(cardMedia).toBeInTheDocument();
  });

  // Test2: caption text content reads correctly
  test("caption content displays correctly", () => {
    render(<ActivityFeed post={fakePost} />);
    const caption = screen.getByTestId("caption");
    expect(caption).toHaveTextContent(
      "Cum consectetur et. At quia sed non accusantium saepe qui cum. Sed est asperiores qui."
    );
  });

  test("card media is shown on the screen on loading", () => {
    const { queryByAltText, queryByTestId } = render(
      <ActivityFeed post={fakePost} />
    );
    expect(queryByAltText("Post Picture")).toBeNull();
    expect(queryByTestId("comment-section")).toBeNull();
  });

  // test("card media is not shown on the screen on loading", () => {
  //   const { queryByTestId } = render(<ActivityFeed post={fakePost} />);
  //   const writeButton = screen.getByTestId("write-test");
  //   expect(queryByTestId("comment-section")).toBeNull();
  //   expect(queryByTestId("form-test")).toBeNull();
  //   expect(writeButton).toBeInTheDocument();
  //   userEvent.click(writeButton);
  //   expect(queryByTestId("comment-section")).toBeInTheDocument();
  // });

  test("comment form renders correctly", () => {
    const { queryByTestId, queryByRole } = render(
      <Context.Provider value={{ fakeLoggedInUser }}>
        <ActivityFeed post={fakePost} />
      </Context.Provider>
    );
    const writeButton = screen.getByTestId("write-test");
    expect(queryByTestId("form-test")).toBeNull();
    expect(writeButton).toBeInTheDocument();
    const childInput = queryByRole("textbox");
    expect(childInput).toBeNull();
    // userEvent.click(writeButton);
    // const commentSection = queryByTestId("comment-section");
    // const childButton = within(commentSection).getByRole("button");
    // const textInput = within(commentSection).getByRole("textbox");
    // expect(childButton).toBeInTheDocument();
    // expect(textInput).toBeInTheDocument();
  });

  // Test3: match snapshots
  it("ActivityFeed renders correctly", () => {
    const tree = renderer
      .create(<ActivityFeed post={fakePost}></ActivityFeed>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
