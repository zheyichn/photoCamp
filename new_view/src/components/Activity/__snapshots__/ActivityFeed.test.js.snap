// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivityFeed Component ActivityFeed renders correctly 1`] = `
<div
  className="MuiGrid-root css-1cvuazx-MuiGrid-root"
>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <div
      className="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
    >
      <div
        className="MuiCardHeader-avatar css-1ssile9-MuiCardHeader-avatar"
      >
        <div
          className="MuiAvatar-root MuiAvatar-circular css-1wlk0hk-MuiAvatar-root"
        >
          <img
            className="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
            src="http://loremflickr.com/640/480/technics"
          />
        </div>
      </div>
      <div
        className="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
      />
    </div>
    <img
      alt="Post Picture"
      className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
      data-testid="test-img"
      height="300px"
      src="https://user-images.githubusercontent.com/93358071/191072941-5fc4a59a-49d0-4b18-80f9-573e38984fa8.jpeg"
    />
    <div
      className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
      data-testid="caption"
    >
      Cum consectetur et. At quia sed non accusantium saepe qui cum. Sed est asperiores qui.
    </div>
    <div
      className="MuiCardActions-root css-1rwjz6-MuiCardActions-root"
    >
      <button
        aria-label="like"
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="FavoriteIcon"
          focusable="false"
          style={
            {
              "color": "grey",
            }
          }
          viewBox="0 0 24 24"
        >
          <path
            d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
          />
        </svg>
      </button>
      <button
        aria-label="comment"
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        data-testid="write-test"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ChatBubbleOutlineRoundedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 4v12H5.17L4 17.17V4h16m0-2H4c-1.1 0-2 .9-2 2v15.59c0 .89 1.08 1.34 1.71.71L6 18h14c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2z"
          />
        </svg>
      </button>
      <button
        aria-label="hide"
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        data-testid="hide-test"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="VisibilityOffIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 7c2.76 0 5 2.24 5 5 0 .65-.13 1.26-.36 1.83l2.92 2.92c1.51-1.26 2.7-2.89 3.43-4.75-1.73-4.39-6-7.5-11-7.5-1.4 0-2.74.25-3.98.7l2.16 2.16C10.74 7.13 11.35 7 12 7zM2 4.27l2.28 2.28.46.46C3.08 8.3 1.78 10.02 1 12c1.73 4.39 6 7.5 11 7.5 1.55 0 3.03-.3 4.38-.84l.42.42L19.73 22 21 20.73 3.27 3 2 4.27zM7.53 9.8l1.55 1.55c-.05.21-.08.43-.08.65 0 1.66 1.34 3 3 3 .22 0 .44-.03.65-.08l1.55 1.55c-.67.33-1.41.53-2.2.53-2.76 0-5-2.24-5-5 0-.79.2-1.53.53-2.2zm4.31-.78 3.15 3.15.02-.16c0-1.66-1.34-3-3-3l-.17.01z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
